### PR TITLE
TiFlash io rate limit configuration.

### PR DESCRIPTION
**Pull request by JinheLin**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### What is changed, added or deleted? (Required)

Add configuration for TiFlash io rate limit.

**This feature is GA in v5.2.**

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Comment by TomShawn** [comment]
@JinheLin Please sign the [Contributor License Agreement](https://cla-assistant.io/pingcap/docs-cn?pullRequest=6820) and involve a technical review. If this PR is v5.2 related, please add the v5.2 label. Thanks!

**Comment by JinheLin** [comment]
> @JinheLin Please sign the [Contributor License Agreement](https://cla-assistant.io/pingcap/docs-cn?pullRequest=6820). If this PR is v5.2 related, please add the v5.2 label. Thanks!

image

I have signed tha CLA.  And I don't have permission to add a lable.

**Comment by TomShawn** [comment]
@JinheLin Please click this button. If it is v5.2-related, please specify in the PR description. Thanks!
image

**Comment by JinheLin** [comment]
> @JinheLin Please click this button. If it is v5.2-related, please specify in the PR description part and I'll add the label. Thanks!
> image

@TomShawn I have clicked this button, and you can see my screenshot above: "You have agreed to the CLA for pingcap/docs-cn".

I have 

> @JinheLin Please click this button. If it is v5.2-related, please specify in the PR description part and I'll add the label. Thanks!
> image

I have clicked this button. You can see my screenshot above: "You have agreed to the CLA for pingcap/docs-cn".

And I have update the PR description part: "This feature is GA in v5.2". Please add a v5.2 label for it. @TomShawn

File: tiflash/tiflash-configuration.md [hunk 1]
```diff
@@ -58,13 +58,7 @@ delta_index_cache_size = 0
...
 ## 存储路径相关配置，从 v4.0.9 开始生效
 [storage]
-    ## [实验特性] 自 v5.0 引入，限制后台任务每秒写入的字节数。目前为实验特性，不推荐在生产环境中使用。
-    ## 以 byte 为单位。目前不支持如 "10GB" 的设置。
-    ## 默认为 0，代表没有限制。
-    ## 该参数主要针对 TiFlash 部署在 AWS EBS (gp2/gp3) 盘时的场景，用于控制后台任务对机器磁盘带宽的占用。
-    ## 提升 TiFlash 查询性能的稳定性。在该场景下推荐配置为磁盘带宽的 50%。
-    ## 其他场景下不建议修改该配置。
-    bg_task_io_rate_limit = 0
+    ## 废弃实验参数：bg_task_io_rate_limit
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:61]:
```suggestion
    ## 该参数从 v5.2.0 开始废弃，请使用 `[storage.io_rate_limit]` 相关配置
    # bg_task_io_rate_limit = 0
```

File: tiflash/tiflash-configuration.md [hunk 2]
```diff
@@ -86,6 +80,33 @@ delta_index_cache_size = 0
     ## storage.latest.dir 存储目录列表中，每个目录的最大可用容量。
     # capacity = [ 10737418240, 10737418240 ]
 
+    [storage.io_rate_limit]
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:84]:
```suggestion
    ## `[storage.io_rate_limit]` 相关配置从 v5.2.0 开始引入 。
    [storage.io_rate_limit]
```

File: tiflash/tiflash-configuration.md [hunk 3]
```diff
@@ -86,6 +81,33 @@ delta_index_cache_size = 0
...
+    ## IO 限流的读写总带宽，单位：Byte。默认值 0，关闭 IO 限流功能。
+    # max_bytes_per_sec = 0
+    ## max_read_bytes_per_sec 和 max_write_bytes_per_sec 的含义和 max_bytes_per_sec 类似。
+    ## 只是这里的读写带宽限制是分开的，适用于一些读写带宽限制是分开计算的云盘，比如 GCP 上的 persistent disk。
+    # max_read_bytes_per_sec = 0
+    # max_write_bytes_per_sec = 0
+
+    ## 下面的参数，一般不需要调整
+    ## TiFlash 内部将 IO 请求分成 4 种类型：前台写、后台写、前台读、后台读。IO 限流初始化时，会根据下面的权重（weight）
+    ## 比例分配带宽。如果 weight 配置为 0，则对应的 IO 操作不会被 IO 限流。
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:95]:
```suggestion
    ## 比例分配带宽。以下默认配置表示每一种流量将获得 1/4 的权重。
    ## 如果 weight 配置为 0，则对应的 IO 操作不会被 IO 限流。
```

File: tiflash/tiflash-configuration.md [hunk 4]
```diff
@@ -86,6 +81,33 @@ delta_index_cache_size = 0
...
+    # foreground_write_weight = 25
+    # background_write_weight = 25
+    # foreground_read_weight = 25
+    # background_read_weight = 25
+    ## TiFlash 支持自动根据当前的 IO 负载情况自动调整各种 IO 类型的限流带宽。
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:101]:
```suggestion
    ## TiFlash 支持自动根据当前的 IO 负载情况自动调整各种 IO 类型的限流带宽，有可能会超过设置的权重。
```

**Comment by TomShawn** [comment]
@flowbehappy Is this PR technically approved?

File: tiflash/tiflash-configuration.md [hunk 5]
```diff
@@ -86,6 +81,34 @@ delta_index_cache_size = 0
     ## storage.latest.dir 存储目录列表中，每个目录的最大可用容量。
     # capacity = [ 10737418240, 10737418240 ]
 
+    [storage.io_rate_limit]
+    ## IO 限流功能主要针对磁盘带宽较小且明确知道磁盘带宽大小的云盘场景，默认关闭。
+    ## IO 限流的读写总带宽，单位：Byte。默认值 0，关闭 IO 限流功能。
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:86]:
```suggestion
    ## I/O 限流功能限制下的读写流量总带宽，单位为 Byte，默认值为 0，即默认关闭 I/O 限流功能。
```

File: tiflash/tiflash-configuration.md [hunk 6]
```diff
@@ -86,6 +81,34 @@ delta_index_cache_size = 0
...
+
+    ## 下面的参数，一般不需要调整
+    ## TiFlash 内部将 IO 请求分成 4 种类型：前台写、后台写、前台读、后台读。IO 限流初始化时，会根据下面的权重（weight）
+    ## 比例分配带宽。以下默认配置表示每一种流量将获得 1/4 的权重。
+    ## 如果 weight 配置为 0，则对应的 IO 操作不会被 IO 限流。
+    # foreground_write_weight = 25
+    # background_write_weight = 25
+    # foreground_read_weight = 25
+    # background_read_weight = 25
+    ## TiFlash 支持自动根据当前的 IO 负载情况自动调整各种 IO 类型的限流带宽，有可能会超过设置的权重。
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:101]:
```suggestion
    ## TiFlash 支持根据当前的 I/O 负载情况自动调整各种 I/O 类型的限流带宽，有可能会超过设置的权重。
```

File: tiflash/tiflash-configuration.md (hunk 2 above)
Comment (TomShawn) [reply tiflash/tiflash-configuration.md:84]: +1

File: tiflash/tiflash-configuration.md [hunk 7]
```diff
@@ -86,6 +81,36 @@ delta_index_cache_size = 0
     ## storage.latest.dir 存储目录列表中，每个目录的最大可用容量。
     # capacity = [ 10737418240, 10737418240 ]
 
+    ## `[storage.io_rate_limit]` 相关配置从 v5.2.0 开始引入 。
+    [storage.io_rate_limit]
+    ## 该配置项是 I/O 限流功能的开关，默认关闭。TiFlash 的 I/O 限流功能适用于磁盘带宽较小且磁盘带宽大小明确的云盘场景。
+    ## I/O 限流功能限制下的读写流量总带宽，单位为 Byte，默认值为 0，即默认关闭 I/O 限流功能。
+    # max_bytes_per_sec = 0
+    ## max_read_bytes_per_sec 和 max_write_bytes_per_sec 的含义和 max_bytes_per_sec 类似，分别指 I/O 限流功能限制下的读流量总带宽和写流量总带宽 。
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:89]: 还需要描述下这两个配置与 max_bytes_per_sec 的优先级关系

Comment (JinheLin) [reply tiflash/tiflash-configuration.md:89]: PTAL: https://github.com/pingcap/docs-cn/pull/6820/commits/eaaf12ddd5b3fc4c1bc21dba43a88a1d3b64e196

**Review by flowbehappy — APPROVED** [review]
LGTM

File: tiflash/tiflash-configuration.md [hunk 8]
```diff
@@ -86,6 +81,37 @@ delta_index_cache_size = 0
...
+    ## TiFlash 支持根据当前的 I/O 负载情况自动调整各种 I/O 类型的限流带宽，有可能会超过设置的权重。
+    ## auto_tune_sec 表示自动调整的执行间隔，单位为秒。设为 0 表示关闭自动调整。
+    # auto_tune_sec = 5
+    ## I/O 带宽占用的百分比的水位线，用于自动调整逻辑，一般不需要调整。
+    # emergency_pct = 96
+    # high_pct = 85
+    # medium_pct = 60
+    ## I/O 带宽调整的限制参数，一般不需要调整
+    # tune_base = 2
+    # min_bytes_per_sec = 2097152
```
[the remark this answers is not in the document]
Comment (flowbehappy) [reply tiflash/tiflash-configuration.md:113]: 这些得删掉好了，不必暴露给用户

```
    ## I/O 带宽占用的百分比的水位线，用于自动调整逻辑，一般不需要调整。
    # emergency_pct = 96
    # high_pct = 85
    # medium_pct = 60
    ## I/O 带宽调整的限制参数，一般不需要调整
    # tune_base = 2
    # min_bytes_per_sec = 2097152
```

File: tiflash/tiflash-configuration.md [hunk 9]
```diff
@@ -86,6 +81,37 @@ delta_index_cache_size = 0
...
+    ## max_read_bytes_per_sec 和 max_write_bytes_per_sec 的含义和 max_bytes_per_sec 类似，分别指 I/O 限流功能限制下的读流量总带宽和写流量总带宽 。
+    ## 分别用两个配置项控制读写带宽限制，适用于一些读写带宽限制分开计算的云盘，例如 GCP 上的 persistent disk。
+    ## 当 max_bytes_per_sec 配置不为 0 时，优先使用 max_bytes_per_sec。
+    # max_read_bytes_per_sec = 0
+    # max_write_bytes_per_sec = 0
+
+    ## 下面的参数用于控制不同 I/O 流量类型分配到的带宽权重，一般不需要调整。
+    ## TiFlash 内部将 I/O 请求分成 4 种类型：前台写、后台写、前台读、后台读。
+    ## I/O 限流初始化时，TiFlash 会根据下面的权重 (weight) 比例分配带宽。
+    ## 以下默认配置表示每一种流量将获得 1/4 的权重。
```
[the remark this answers is not in the document]
Comment (JinheLin) [reply tiflash/tiflash-configuration.md:98]: 以下默认配置表示每一种流量将获得 25 / (25 + 25 + 25 + 25) = 25% 的权重。

@SE-Bin 这样的描述有没有问题？

[the remark this answers is not in the document]
Comment (JinheLin) [reply tiflash/tiflash-configuration.md:98]: 以下默认配置表示每一种流量将获得 25 / (25 + 25 + 25 + 25) = 25% 的权重。

@SE-Bin 这样的描述有没有问题？

File: tiflash/tiflash-configuration.md (hunk 7 above)
Comment (flowbehappy) [reply tiflash/tiflash-configuration.md:89]:
```suggestion
    ## max_read_bytes_per_sec 和 max_write_bytes_per_sec 的含义和 max_bytes_per_sec 类似，分别指 I/O 限流功能限制下的读流量总带宽和写流量总带宽。
```

File: tiflash/tiflash-configuration.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (SE-Bin) [reply tiflash/tiflash-configuration.md:98]: 我觉得文字描述直接用 25%  就可以了。跟下面的配置项的 25 保持一致就 OK 。

File: tiflash/tiflash-configuration.md [hunk 10]
```diff
@@ -86,6 +81,30 @@ delta_index_cache_size = 0
     ## storage.latest.dir 存储目录列表中，每个目录的最大可用容量。
     # capacity = [ 10737418240, 10737418240 ]
 
+    ## `[storage.io_rate_limit]` 相关配置从 v5.2.0 开始引入 。
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:84]:
```suggestion
    ## `[storage.io_rate_limit]` 相关配置从 v5.2.0 开始引入。
```

File: tiflash/tiflash-configuration.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (flowbehappy) [reply tiflash/tiflash-configuration.md:98]:
```suggestion
    ## 以下默认配置表示每一种流量将获得 25% 的权重。
```

File: tiflash/tiflash-configuration.md [hunk 11]
```diff
@@ -86,6 +81,30 @@ delta_index_cache_size = 0
...
+    ## max_read_bytes_per_sec 和 max_write_bytes_per_sec 的含义和 max_bytes_per_sec 类似，分别指 I/O 限流功能限制下的读流量总带宽和写流量总带宽。
+    ## 分别用两个配置项控制读写带宽限制，适用于一些读写带宽限制分开计算的云盘，例如 GCP 上的 persistent disk。
+    ## 当 max_bytes_per_sec 配置不为 0 时，优先使用 max_bytes_per_sec。
+    # max_read_bytes_per_sec = 0
+    # max_write_bytes_per_sec = 0
+
+    ## 下面的参数用于控制不同 I/O 流量类型分配到的带宽权重，一般不需要调整。
+    ## TiFlash 内部将 I/O 请求分成 4 种类型：前台写、后台写、前台读、后台读。
+    ## I/O 限流初始化时，TiFlash 会根据下面的权重 (weight) 比例分配带宽。
+    ## 以下默认配置表示每一种流量将获得 25% 的权重。
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:98]:
```suggestion
    ## 以下默认配置表示每一种流量将获得 25 / (25 + 25 + 25 + 25) = 25% 的权重。
```

**Comment by SE-Bin** [comment]
LGTM

File: tiflash/tiflash-configuration.md [hunk 12]
```diff
@@ -86,6 +81,30 @@ delta_index_cache_size = 0
     ## storage.latest.dir 存储目录列表中，每个目录的最大可用容量。
     # capacity = [ 10737418240, 10737418240 ]
 
+    ## `[storage.io_rate_limit]` 相关配置从 v5.2.0 开始引入。
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:84]:
```suggestion
    ## [storage.io_rate_limit] 相关配置从 v5.2.0 开始引入。
```

**Review by TomShawn — APPROVED** [review]

**Pull request merged**